geode-core/src/test/java/org/apache/geode/distributed/internal/membership/MembershipDependenciesJUnitTest.java

```diff
@@ -0,0 +1,275 @@
+import org.apache.geode.internal.logging.log4j.AlertAppender;
+import org.apache.geode.internal.logging.log4j.LogMarker;
+import org.apache.geode.internal.net.SocketCreator;
+import org.apache.geode.internal.net.SocketCreatorFactory;
+import org.apache.geode.internal.security.SecurableCommunicationChannel;
+import org.apache.geode.internal.shared.StringPrintWriter;
+import org.apache.geode.internal.tcp.ConnectExceptions;
+import org.apache.geode.internal.util.Breadcrumbs;
+import org.apache.geode.internal.util.JavaWorkarounds;
+
+@RunWith(ArchUnitRunner.class)
+@AnalyzeClasses(packages = "org.apache.geode.distributed.internal.membership..")
+public class MembershipDependenciesJUnitTest {
+
+  /*
+   * This test verifies that the membership component (which is currently made up of classes
+   * inside the geode-core module, but which may someday reside in a separate module)
+   * depends only on packages within itself (within the membership component) or packages
+   * outside apache.geode.
+   *
+   * For purposes of this test, classes in the membership...adapter package are not considered.
+   * They will eventually become part of geode-core.
+   *
+   * While this rule is ignored, comment-out the ignore annotation to run it periodically to get
+   * the current count of deviations.
+   */
+  // TODO: remove ignore once membershipDoesntDependOnCoreProvisional matches this rule exactly
+  @ArchIgnore
+  @ArchTest
+  public static final ArchRule membershipDoesntDependOnCore = classes()
+      .that()
+      .resideInAPackage("org.apache.geode.distributed.internal.membership..")
+      .and()
+      .resideOutsideOfPackage("org.apache.geode.distributed.internal.membership.adapter..")
+      .should()
+      .onlyDependOnClassesThat(
+          resideInAPackage("org.apache.geode.distributed.internal.membership..")
+              .or(not(resideInAPackage("org.apache.geode.."))));
+
+  /*
+   * This test is a work-in-progress. It starts from the membershipDoesntDependOnCore rule
+   * and adds deviations. Each deviation has a comment like TODO:...
+   * Those deviations comprise a to do list for the membership team as it modularizes
+   * the membership component--severing its dependency on the geode-core component.
+   */
+  @ArchTest
+  public static final ArchRule membershipDoesntDependOnCoreProvisional = classes()
+      .that()
+      .resideInAPackage("org.apache.geode.distributed.internal.membership..")
+      .and()
+      .resideOutsideOfPackage("org.apache.geode.distributed.internal.membership.adapter..")
+      .and()
+
+      // TODO: InternalDistributedMember needs to move out of the package
+      .areNotAssignableFrom(InternalDistributedMember.class)
+
+
+      .should()
+      .onlyDependOnClassesThat(
+          resideInAPackage("org.apache.geode.distributed.internal.membership..")
+
+              .or(not(resideInAPackage("org.apache.geode..")))
+
+              // TODO: Create a new stats interface for membership
```

Comment:
Belated booooooooooooo!

I hope you haven't opened JIRA tickets for these `TODO`s, because no one is looking here for work to do.